src/equations/acoustic_perturbation_2d.jl

```diff
@@ -104,51 +91,198 @@ function source_terms_convergence_test(u, x, t, equations::AcousticPerturbationE
+                           u_inner[7])
+    else # Wall
+      u_boundary = SVector(u_inner[1], -u_inner[2], u_inner[3], u_inner[4], u_inner[5], u_inner[6],
+                           u_inner[7])
+    end
+  else
+    u_boundary = SVector(0.0, 0.0, 0.0, u_inner[4], u_inner[5], u_inner[6], u_inner[7])
+  end
+
+  # Calculate boundary flux
+  if direction in (2, 4) # u_inner is "left" of boundary, u_boundary is "right" of boundary
+    flux = surface_flux_function(u_inner, u_boundary, orientation, equations)
+  else # u_boundary is "left" of boundary, u_inner is "right" of boundary
+    flux = surface_flux_function(u_boundary, u_inner, orientation, equations)
+  end
+
+  return flux
 end
 
 
 # Calculate 1D flux for a single point
 @inline function calcflux(u, orientation, equations::AcousticPerturbationEquations2D)
-  v1_prime, v2_prime, p_prime = u
-  @unpack v_mean, rho_mean, c_mean = equations
+  v1_prime, v2_prime, p_prime, v1_mean, v2_mean, c_mean, rho_mean = u
 
+  # Calculate flux for conservative state variables
   if orientation == 1
-    f1 = v_mean[1] * v1_prime + v_mean[2] * v2_prime + p_prime / rho_mean
+    f1 = v1_mean * v1_prime + v2_mean * v2_prime + p_prime / rho_mean
     f2 = zero(eltype(u))
-    f3 = c_mean^2 * rho_mean * v1_prime + v_mean[orientation] * p_prime
+    f3 = c_mean^2 * rho_mean * v1_prime + v1_mean * p_prime
   else
     f1 = zero(eltype(u))
-    f2 = v_mean[1] * v1_prime + v_mean[2] * v2_prime + p_prime / rho_mean
-    f3 = c_mean^2 * rho_mean * v2_prime + v_mean[orientation] * p_prime
+    f2 = v1_mean * v1_prime + v2_mean * v2_prime + p_prime / rho_mean
+    f3 = c_mean^2 * rho_mean * v2_prime + v2_mean * p_prime
   end
 
-  return SVector(f1, f2, f3)
+  # The rest of the state variables are actually variable coefficients, hence the flux should be
+  # zero. See https://github.com/trixi-framework/Trixi.jl/issues/358#issuecomment-784828762
+  # for details.
+  f4 = f5 = f6 = f7 = zero(eltype(u))
+
+  return SVector(f1, f2, f3, f4, f5, f6, f7)
 end
 
 
 function flux_lax_friedrichs(u_ll, u_rr, orientation, equations::AcousticPerturbationEquations2D)
   f_ll = calcflux(u_ll, orientation, equations)
   f_rr = calcflux(u_rr, orientation, equations)
 
-  v_ll = u_ll[orientation] + equations.v_mean[orientation]
-  v_rr = u_rr[orientation] + equations.v_mean[orientation]
-  c0 = equations.c_mean
-  speed = max(abs(v_ll), abs(v_rr)) + c0
+  # Calculate v = v_prime + v_mean
+  v_ll = u_ll[orientation] + u_ll[orientation + 3]
+  v_rr = u_rr[orientation] + u_rr[orientation + 3]
+
+  c_mean_ll = u_ll[6]
+  c_mean_rr = u_rr[6]
+
+  speed = max(abs(v_ll), abs(v_rr)) + max(c_mean_ll, c_mean_rr)
 
   return 0.5 * ( (f_ll + f_rr) - speed * (u_rr - u_ll) )
 end
 
 
-@inline have_constant_speed(::AcousticPerturbationEquations2D) = Val(true)
+@inline have_constant_speed(::AcousticPerturbationEquations2D) = Val(false)
+
+@inline function max_abs_speeds(u, equations::AcousticPerturbationEquations2D)
```

Comment:
This is problematic (slightly): It means we have to recompute the time step size in each time step, which is really not necessary. Ideally, we could have another type trait (in addition to `have_constant_speed`) that asks whether the speed is constant in time. Or maybe a third option for `have_constant_speed` that returns `Val(:time)` that indicates that the speeds are only constant in time, but not in space, such that the semidiscretization has to store it.

@ranocha Do you have an idea how to implement this in a sensible manner without too much hassle? Otherwise we can just live with the overhead for now of having to recompute the time step size in each iteration.

Reply:
I see your point, but until #215 is resolved the step size will be recalculated in each step anyway, if I understand correctly. In that case, does it make sense to leave this as is and track this in a separate issue?

Reply:
Agreed - just leave it as it is.